.github/actions/sonar/action.yml

```
@@ -0,0 +1,259 @@
+#
+# Copyright 2013-2024 the original author or authors from the JHipster project.
+#
+# This file is part of the JHipster project, see https://www.jhipster.tech/
+# for more information.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+name: 'SonarQube PR Analysis'
+description: 'A GitHub Action to perform SonarQube analysis on a PR with caching and metrics retrieval.'
+
+inputs:
+  sonar-project-key:
+    description: 'SonarQube project key'
+    required: true
+  workdir:
+    description: 'Working directory for the action'
+    required: true
+  docker-compose-file:
+    description: 'Path to the Docker Compose file'
+    required: true
+  comment-token:
+    description: 'GitHub token for commenting on the PR'
+    required: true
+
+runs:
+  using: composite
+  steps:
+    # Cache the main branch project for future PR analysis
+    - name: 'Cache project on main branch'
+      if: github.ref == 'refs/heads/main'
+      shell: bash
+      run: |
+        mkdir -p ./${{ inputs.sonar-project-key }}
+        cp -r ${{ inputs.workdir }}/. ./${{ inputs.sonar-project-key }}/
```

Cleanup and reduce application size to be cached.
```suggestion
        cp -r ${{ inputs.workdir }}/. ./${{ inputs.sonar-project-key }}/
        rm -rf ${{ inputs.sonar-project-key }}/node_modules \
            ${{ inputs.sonar-project-key }}/package-lock.json \
            ${{ inputs.sonar-project-key }}/target/node \
            ${{ inputs.sonar-project-key }}/target/*.original \
            ${{ inputs.sonar-project-key }}/target/*.jar
```